### PR TITLE
Support for @Provider discovery in MP server (#1432)

### PR DESCRIPTION
**Pull request by aseovic**
Adds support for manual registration of providers (via Builder), as well as automatic discovery and registration of CDI beans with `@Provider` annotation

### CODE REVIEW

File: microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018-2020 Oracle and/or its affiliates.
```
Comment (tomas-langer) [inline microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java]: Copyright years should be comma separated in Helido.

File: microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java [hunk 2]
```diff
@@ -315,7 +320,7 @@ private void resourcesFromContainer() {
         private SeContainer createContainer(ClassLoader classLoader) {
             // not in CDI
             Weld initializer = new Weld();
-            initializer.addBeanDefiningAnnotations(Path.class);
+            initializer.addBeanDefiningAnnotations(Path.class, Provider.class);
```
Comment (tomas-langer) [inline microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java]: As you said in chat, we want to load only providers that are annotated using a bean defining annotation. This change modifies that behavior by making all of the having a bean defining annotation.

File: microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java (hunk 1 above)
Comment (aseovic) [reply microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java]: Hmm... I'm pretty sure I've seen dash used somewhere, so we are definitely not consistent.

File: microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java (hunk 2 above)
Comment (aseovic) [reply microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java]: I just wanted to make sure it works exactly the way resources work. It is my understanding that this code ensures that any resources (and now providers) that have been explicitly registered via Builder are also registered with the CDI container. 

I think that makes sense, and is separate from CDI discovery, which is done by the extension and expects providers to be proper managed beans.

File: microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java (hunk 1 above)
Comment (aseovic) [reply microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java]: Fixed.

File: microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java (hunk 2 above)
Comment (tomas-langer) [reply microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java]: The problem is that this is a backward incompatible change and may have unexpected behavior.
If you add `@ApplicationScoped` to a class annotated as `@Provider`, it will be automatically picked up by your changes.
Adding `Provider` as a bean defining annotation is a step too far, as any provider in the code base would get added automatically, and that may not be desired. I would keep the original agreed approach that only beans are added automatically

Comment (spericas) [reply microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java]: Perhaps this is something we can do in 2.0 where the scanning code is already incompatible

Comment (aseovic) [reply microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java]: But that's the thing -- it doesn't, just like the resources annotated with`@Path` do not get automatically picked **unless** they are also annotated with `@ApplicationScoped` or `@RequestScoped`.

The only thing the code above does is **allow** you to register classes that do not have any standard bean defining annotations with the bean manager, which happens a little bit later in the code:

```
            // add resource classes explicitly configured without CDI annotations
            this.resourceClasses.stream()
                    .filter(this::notACdiBean)
                    .forEach(initializer::addBeanClasses);

            // add provider classes explicitly configured without CDI annotations
            this.providerClasses.stream()
                    .filter(this::notACdiBean)
                    .forEach(initializer::addBeanClasses);
```
I've tried running my demo app using this implementation, and my filter is only discovered and registered automatically if I add `@ApplicationScoped` annotation as well. None of the other `@Provider` classes get picked up automatically, even though I depend on MP bundle and am bringing in a ton of dependencies.

The bottom line is that because Jersey treats resources and providers the same (they are all returned from the same `Application.getClasses` method), we should treat them the same as well and allow for one what we already allow for the other (in this case, registration of explicitly added resources and providers that are **not** proper managed beans with bean manager). I thought about not even separating them out and using a single list to collect all resource and provider classes, but decided against it as we have a few places where we log warning if there are no resources defined for the app, so having them separate is helpful.

Comment (tomas-langer) [reply microprofile/server/src/main/java/io/helidon/microprofile/server/Server.java]: I just realized this is the 1.4 branch, where we handle resources differently from 2.0. On `master` some of these assumptions are no longer valid. Sorry about that.
We probably can keep the same function for 2.0 when it comes to `@Provider`, though it does work differently for `@Path`

**Comment by aseovic** [comment]
I've added the tests for new functionality, as we discussed, and hit the bug with MBR/MBW generic type argument processing when Weld proxies are used. I'll open a bug against Jersey for that, so we can track it.

Apart from that, I've changed some of the server semantics a bit in order to be able to mix explicit applications with auto-discovered resources and providers (implicit application).

Basically, you can now register one or more `Application` classes explicitly, just like before, but if you have explicitly registered or auto-discovered resources and providers in addition to those, they will be packaged into an implicit/default application and added to the `applications` list. That way it's not "either/or" as it is at the moment -- you can have both at the same time.

The main reason I did it was so I can have both in the test project, but I do think it makes sense in general -- it allows you to implement your own app/service using auto-discoverable resources and providers, but also add explicit apps defined elsewhere.

I've also changed `ServerCdiExtension` to support discovery of `@RequestScoped` root resources (which is an issue I've hit before but didn't know why they are not being picked up automatically), and changed it discovers everything that's discoverable, not either applications or resources/providers.

Anyway, let's discuss tomorrow and decide how to proceed.

**Comment by aseovic** [comment]
Opened https://github.com/eclipse-ee4j/jersey/issues/4400 to track the issue with CDI-managed MessageBodyReaders/MessageBodyWriters.

**Comment by tomas-langer** [comment]
The combination of discovered resource classes and application is intentionally not possible. You can add an auto-discovered application that explicitly returns resource and provider classes and combine it with registered applications
You cannot have registered application classes and combine them with discovered resources, as that may create unexpected endpoints.

**Comment by tomas-langer** [comment]
Do not merge this PR please, see #1448
